### ISSUE TITLE
pwt/pyret-parser-tests breaks drag&drop

### ISSUE THREAD

**Issue by justinpombrio** [bootstrapworld/codemirror-blocks#232]
Something important broke in this branch, taking out drag&drop with it. The error I'm seeing is the failure of `assert(!newPtr.done)` in ast-patch. (That's a correct assertion, though.)

**Comment by schanzer** [comment]
Do you have a sense for which commit broke it? Or has it always been broken in this branch?

**Comment by prestontw** [comment]
We're guessing it's the hash commits---that's the only thing that touches WeScheme off the top of my head...

**Comment by schanzer** [comment]
That certainly makes sense. It looks like you’re hashing on the object itself, rather than the (string) node type?

**Comment by schanzer** [comment]
Ah, I see the issue. You've moved hashing to the super constructor (`ASTNode`) for each node type, but that hashing requires calling the `...this.children()` iterator in the super constructor.  That iterator assumes that `this` has fields set for each of the `keys`...but by definition, calling the super constructor has to happen _before_ setting any fields of `this`.

I think the way forward is to define a `setHash` function on the `ASTNode`, and then override that for special nodes like `Literal`. Then, in each of the nodes, add `this.hash = this.setHash()` as the last line in the constructor. You don't get any LOC savings, but at least you get the logic cleanup that you proposed here.

**Comment by schanzer** [comment]
I've implement this on `master` as 5053609

**Comment by prestontw** [comment]
I'll implement this for Pyret nodes soon. In the meantime, drag-and-drop works for WeScheme nodes again.

**Comment by schanzer** [comment]
Is this resolved? Or waiting on a merge into master?

**Comment by prestontw** [comment]
This is resolved! I've merged master into ptw/pyret-parser-tests
